### DISCUSSION TITLE
Node-level masking and prediction in Disjoint Mode

### DISCUSSION THREAD

**Discussion by frakt0x90** [danielegrattarola/spektral#310] [Q&A]
I have a bunch of  graphs (< 100 nodes, ~800 edges) with no node or edge features, identical adjacency matrices, but different labels. Given a subset of labels on a graph, there's a deterministic way to determine the rest so I was hoping to train a GNN to do this. After reading the docs it seems like disjoint mode makes the most sense as I can treat them like a single big graph and mask some nodes for testing. My question revolves around the best way to do that.
 
All the examples I found for disjoint mode are doing graph rather than node classification. I currently have a Dataset that inherits from spektral.data.Dataset and implements read() to return a list of spektral.data.Graph objects that correspond to my graphs. Now to split this data into test and train, do I simply create:

`train = DisjointLoader(dataset[:train_graphs], node_level=True)`
`test = DisjointLoader(dataset[train_graphs:], node_level=True)`

but how do I mask a portion of the node labels for training and provide the ground truth?

Or would it be easier to just make the single disjoint structure manually via `spektral.data.utils.to_disjoint()`, modify my labels to match the expanded graph and then use the SingleLoader following the [citation gcn example](https://github.com/danielegrattarola/spektral/blob/master/examples/node_prediction/citation_gcn.py) since there aren't any node or edge features and the adjacency matrices are identical?

Thanks for your help and for making this cool project!

**Comment by danielegrattarola** [comment]
Sorry, this slipped through my inbox. 

You have a couple of ways to do what you want:

- Use DisjointLoader with `node_level=True` and augment the labels with the train/val/test mask. You can then write your own training loop and retrieve the mask from the label once you get a batch. This has the advantage that you will be able to shuffle the graphs at every pass on the dataset.
- Do the same, but with a MixedLoader (saves memory by not duplicating the adjacency matrix)
- Do as you said, using a SingleLoader after having pre-batched the training/test/val graphs. This will let you use the `model.fit` approach. 

Cheers
